### DISCUSSION TITLE
Tools API for your Copilot Extensions

### DISCUSSION THREAD

**Discussion by isidorn** [microsoft/vscode-discussions#2411] [Announcements]
We have finalized the Tools API 🚀 
Docs can be found here https://code.visualstudio.com/api/extension-guides/tools

Try out the API and let us know what you think about it. 

Check out this sample:
[https://github.com/microsoft/vscode-extension-samples/blob/main/chat-sample/README.md](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fvscode-extension-samples%2Fblob%2Fmain%2Fchat-sample%2FREADME.md&data=05%7C02%7CIsidor.Nikolic%40microsoft.com%7Cfb88490c209248bb316e08dd6255896e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638774841714147884%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=6CKuUEzGvwADAeuIjpP%2Fl7kyo0Q%2FHOa5BwI9nGG4dRI%3D&reserved=0)

Add the following to package.json:
"enabledApiProposals": ["languageModelToolsForAgent”],

In langaugeModelTools in package.json, ensure you add the vscode_editing tag:

You then should be able to go to agent mode and your tool can be invoked from a user prompt:

This should be enough to get you started and show what’s possible.

We are still trying to understand what gaps we might have in this approach and what adjustments/affordances we may need to have to make the experience work. As you start building out a POC with some developer scenarios, please report back if there’s anything you feel is required to land your testing scenarios.

Thank you 🙏

**Comment by isidorn** [comment]
We have launched agent mode in preview https://code.visualstudio.com/docs/copilot/copilot-edits#_use-agent-mode-preview
And in the next couple of months we will explore Agent calling tools contributed via the language model tools api.

**Reply by doggy8088** [reply]
Are there any currently released extensions that have implemented the Tools API? I want to try and see how to call these tools in Agent Mode.

**Reply by isidorn** [reply]
Agent Mode is still not using tools from extensions. So unfortunetly there is nothing for you to try out right now. As soon as we have something I will update this discussion.
Thank you for your interest

**Reply by doggy8088** [reply]
So now, in Agent mode, besides the built-in tools, the only custom external tool that can be invoked is the command-line tool in the Terminal, right?

**Reply by doggy8088** [reply]
Will it be possible in the future to invoke any VS Code Commands through Agent Mode?

**Reply by isidorn** [reply]
No.

**Reply by lemmy** [reply]
@isidorn, could you clarify why VSCode's agent mode won't support invoking existing VSCode commands, even if it's just a whitelisted subset? As an extension developer, it seems redundant to have to reimplement each command using `vscode.LanguageModelTool`, especially when those commands are already defined and available in my extension.

**Reply by isidorn** [reply]
Because language model needs nice description of functionality - to decide if it should call it or not. commands do not have that by default.

So please do contribute tools! Thanks 🙏

**Reply by lemmy** [reply]
Here's my follow-up at https://github.com/microsoft/vscode-discussions/discussions/2411#discussioncomment-12938867 ;-)

**Comment by ArtemGovorov** [comment]
@isidorn Hey, coming here from the https://code.visualstudio.com/blogs/2025/02/24/introducing-copilot-agent-mode post:

> Exploring [tool extensibility](https://code.visualstudio.com/api/extension-guides/tools) and [MCP servers](https://github.com/modelcontextprotocol/servers) as tools for agent mode. If you have extension ideas [tell us](https://github.com/microsoft/vscode-discussions/discussions/2411)

We have recently added support for Copilot chat in [Wallaby](https://wallabyjs.com/blog/wallaby-ai.html), also working on adding support in [Console Ninja](https://console-ninja.com/). We have implemented a few `vscode.LanguageModelTool` tools for the chat. Our tools allow LLM to retrieve code execution paths and runtime values for better runtime errors investigation. It would be great to be able to expose those tools to Copilot Edits and Agent as well. 

In terms of tool contribution, some form of tool registration for the Edits/Agent from within a running extension would be great, because it would be much easier to get access to the full context of the extension's state from the tool. The current way to expose tools to chat via a chat participant `@...` mention seems to be not very flexible neither discoverable for the potential variety of tools that may be contributed. 

MCP server support would also work for us in some way, but I guess extensions may need more plumbing to be able to connect the MCP server to the running extension itself. With MCP, it would be great to at least have [MCP roots support](https://modelcontextprotocol.io/docs/concepts/roots), so that the tool can know about the context it is running in, should it somehow need to connect/talk to a running extension to get some state or invoke some code of the extension.

**Comment by vnijs** [comment]
Just saw [this video](https://www.youtube.com/watch?v=CDjjaTALI68) from the LangChain folks. MCP connections are possible in Cursor, Windsurf, and Claude Desktop. I like Copilot Edits and plan to use it with my students. Adding the option to connect MCP to VS Code Copilot would be very very nice. Thanks for your consideration.

**Reply by isidorn** [reply]
Thanks - we are adding MCP tool support to Copilot Agent mode. You can already try it out with VS Code Insiders. And should be in VS Code stable in the next couple of weeks.

**Comment by mohamedmansour** [comment]
Hello! I work on the Chromium code base which is massive, and I Would like to write an agent that uses `git grep` to search for code / files / etc, how can I make that happen? `git grep` is fast to search and I would like it to take my prompt and do a bunch of git grep to figure out which one it should include in my context (to fill the context window)

**Reply by isidorn** [reply]
Write an extension, that contributes a `git grep` tool - then agent should use the tool your extension is contributing.

You can extend agent mode today in VS Code Insiders with the LanguageModelTools API. Check out this sample:
[https://github.com/microsoft/vscode-extension-samples/blob/main/chat-sample/README.md](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fvscode-extension-samples%2Fblob%2Fmain%2Fchat-sample%2FREADME.md&data=05%7C02%7CIsidor.Nikolic%40microsoft.com%7Cfb88490c209248bb316e08dd6255896e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638774841714147884%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=6CKuUEzGvwADAeuIjpP%2Fl7kyo0Q%2FHOa5BwI9nGG4dRI%3D&reserved=0)

Add the following to package.json:
"enabledApiProposals": ["languageModelToolsForAgent”],

In langaugeModelTools in package.json, ensure you add the vscode_editing tag:

You then should be able to go to agent mode and your tool can be invoked from a user prompt:

This should be enough to get you started and show what’s possible.

We are still trying to understand what gaps we might have in this approach and what adjustments/affordances we may need to have to make the experience work. As you start building out a POC with some developer scenarios, please report back if there’s anything you feel is required to land your testing scenarios.

**Reply by ropevp** [reply]
What tool, maybe an MCP tool would you recommend using like obsidian, to link into the codebase, to draw on ? is there a local rag available for this any plans to extend the brain of copilot with local knowledge - maybe a better way to go about.
MCP Marketplace like Cline that would get people excited - but it must get EVERYTHING and amazing support.

**Comment by aymuos15** [comment]
Hello! Thank you so much for this!

How do I allow the agent to use the terminal and run stuff (or force it to run only one command if possible) without asking for my permission each time?

**Reply by isidorn** [reply]
Not possible this milestone. We are adding it in April.

**Reply by aymuos15** [reply]
Thanks for the response! Eagerly waiting :)

**Reply by isidorn** [reply]
fyi @Tyriar

**Comment by gokul427** [comment]
Apart from tools, do we get other primitives like resources and prompts for agent mode mcp?

**Reply by isidorn** [reply]
This is on the plan but has not happened yet.

fyi @digitarald

**Comment by lemmy** [comment]
I've integrated a few languageModelTools into our extension, and the models can now successfully invoke them. However, the `modelDescription` field is proving too limited and cumbersome for providing the level of guidance needed. We've found that, to ensure the LLMs use these tools correctly, we also need to include more comprehensive instructions—specifically, through a `.github/copilot-instructions.md` file.
What is unclear is how to include or bundle this `copilot-instructions.md` file as part of *our* VSCode extension. Right now, the file lives in the user's workspace, which is outside the control of our extension. Is there a way to package it with the extension so we can manage and distribute it reliably, or are we missing something and there is a better alternative to `copilot-instructions.md`.

**Reply by isidorn** [reply]
We would love if the LLM can use the tools correctly without extensions modifying the `copilot-instructions.md` file.
Did you try fine-tuning your tool description to make sure the LLM uses it correctly? Are you using Sonnet or 4.1, since we notice 4o is not good at tool calling.

**Reply by lemmy** [reply]
As extension developers, we don't always know which LLMs our users will be working with. 

Right now, the `modelDescription` field in `package.json` is too limited. Could you introduce something like a `modelDescriptionHook` instead? This hook could be called whenever an LLM requests a tool description, i.e., what's `modelDescription` today, allowing extension developers to generate a tool description dynamically. This would make it possible to tailor the description based on the specific LLM in use and even adjust it according to the tool’s capabilities.

**Reply by isidorn** [reply]
This is a fair feature request - feel free to open one https://github.com/microsoft/vscode/issues

**Reply by lemmy** [reply]
https://github.com/microsoft/vscode/issues/247426

**Comment by lemmy** [comment]
A VSCode extension exposes an underlying (deterministic) tool, such as a compiler, within the VSCode. One approach to integrating this functionality into an agentic workflow is through `languageModelTools`. A different architecture uses MCP  to expose the compiler. Given that both `languageModelTools` and MCP aim to expose tools to LLMs, what is the long-term vision for each? Are they going to coexist, diverge, or converge?  Currently, MCP offers the advantage of being independent of VSCode, making it easier to test agentic workflows. This is particularly relevant because testing agentic workflows is not yet fully available within the VSCode + `languageModelTools`. On the other hand, `languageModelTools` offers a much more user-friendly experience. Users can access tools without needing to run separate MCP servers, making it more convenient for those working entirely within VSCode.

**Comment by netbrah** [comment]
This might be of some interest to folks:

https://github.com/sehejjain/Language-Server-MCP-Bridge

I am also working on a fork of Claude-context to leverage language client passing to boost the existing AST parser that's in place
